### ISSUE TITLE
Failing to parse dates

### ISSUE THREAD

**Issue by BenjaminRH** [c3js/c3#491]
When I use the `timeseries` type, I keep getting the "Failed to parse x to Date object" error. On dates like '2014-08-12T00:00:00.000Z', which when I try to make with `new Date` in the console, work just fine. Any idea why this is happening?

**Comment by fibsifan** [comment]
I still havethis issue. I have set xFormat correctly and use an URL to pull JSON.
code looks like this:
```javascript
data: {
	url: './json.php',
	mimeType: 'json',
	keys: {
		x: 'TIMESTAMP',
		xFormat: '%Y-%m-%d %H:%M:%S',
		value: ['DOWNLOAD']
	}
},
axis: {
	x: {
		type: 'timeseries',
		tick: {
			format: '%Y-%m-%d %H:%M:%S'
		}
	}
}
```
Am I missing something?

**Comment by JiaJian** [comment]
@fibsifan would it be possible to see a sample of your JSON API response? Your response's timestamp needs to follow `data.keys.xFormat` (i.e. `%Y-%m-%d %H:%M:%S`) rather than `x.tick.format`.

**Comment by fibsifan** [comment]
Ah yes, sorry I forgot. It looks like this:
```json
[{"TIMESTAMP":"2015-03-03 23:58:41","DOWNLOAD":"31.20"},{"TIMESTAMP":"2015-03-04 00:07:16","DOWNLOAD":"29.96"}]
```

**Comment by jonnywyatt** [comment]
@fibsifan - try a simple transform to get your JSON data into columns and use that API instead of loading JSON directly and using 'keys'. I'm guessing (but would need verifying) that this might work.

**Comment by fibsifan** [comment]
Just in case someone stumbles upon this: for me the trick was to put xFormat outside AND inside keys:
```javascript
data: {
    url: './json.php',
    mimeType: 'json',
    keys: {
        x: 'TIMESTAMP',
        xFormat: '%Y-%m-%d %H:%M:%S',
        value: ['DOWNLOAD']
    },
    xFormat: '%Y-%m-%d %H:%M:%S'
},
axis: {
    x: {
        type: 'timeseries',
        tick: {
            format: '%Y-%m-%d %H:%M:%S'
        }
    }
}
```

**Comment by just6700** [comment]
@fibsifan - I got the same problem and your trick worked for me! Thanks!

**Comment by dmiddlecamp** [comment]
Just hit this again today, it'd be nice if this worked without a workaround.

Especially since Date.parse("_INSERT_ISO8601_DATE_HERE") works as expected

**Comment by aubricus** [comment]
Just ran into this myself. The format suggested early on does not actually match the date format requested in the original post. 

Got lucky and found a helpful hint [here in the D3 wiki about ISO formatting.](https://github.com/mbostock/d3/wiki/Time-Formatting#format_iso)

*This worked for me:*

- use xFormat prop as recommended
- use this format value: `%Y-%m-%dT%H:%M:%S.%LZ`
- [Here's a JSFiddle for proof](https://jsfiddle.net/aubricus/r2aj35sd/)

**Comment by TheSeldonPlan** [comment]
@aubricus Thanks for posting the link to D3 ISO formatting, :+1:

**Comment by davegonzalez** [comment]
Hi there! 

In c3.js v4.10 I was able to pass moment.js objects into the `x` column and the dates were parsed without any problems. In v4.11rc4, I now run into `Failed to parse x '1461729600000' to Date object` errors.

This occurs even when I set the `xFormat` on the data object. My full object looks like this once it's built:

```
data: {
  columns: [
    ['Total Plays', '4', '2', '5', '3', '4', '4', '3', '2'],
    ['Total Finishes', '1', '2', '3', '0', '2', '0', '2', '2'],
    ['x',  '2016-04-20T04:00:00.000Z', '2016-04-21T04:00:00.000Z', '2016-04-22T04:00:00.000Z', '2016-04-23T04:00:00.000Z', '2016-04-24T04:00:00.000Z', '2016-04-25T04:00:00.000Z', '2016-04-26T04:00:00.000Z', '2016-04-27T04:00:00.000Z']
  ],
  types: {
    'Total Plays': 'spline', 
    'Total Finishes': 'spline'
  },
  x: 'x',
  xFormat: '%Y-%m-%dT%H:%M:%S.%LZ'
}
```

The date object in `c3_chart_internal_fn.parseDate` looks like this:
![screen shot 2016-04-27 at 10 13 14 am](https://cloud.githubusercontent.com/assets/1012987/14855166/b63374b6-0c60-11e6-88fd-4ece9135a64f.png)

**Comment by dabada83** [comment]
I have the same issue that @davegonzalez with 0.4.11 version

**Comment by misham** [comment]
Solved this by changing the format I'm feeding the dates into C3.js.  Have the API send the data in the following format with `%Y-%m-%d %H:%M %Z`  or Rails formatting: `%Y-%m-%d %H:%M %z`.

Then set your `data.xFormat = '%Y-%m-%d %H:%M %Z`

I did not need `axis.x.localtime`, the system just converted the ticks to the right timezone.

For me, the reason it was failing is that Rails sends timestamps in UTC by default.  When it does that, it just assumes the client will know it's already in UTC and not try to parse the timezone.

So the example timestamps above (`2016-04-21T04:00:00.000Z`) do not actually include a timezone, just the date with time, including milliseconds.  The __Z__ in there is supposed to signify that it's UTC.  This is the default serializer btw.  So if you're going to send timestamps from Rails, you'll have to overwrite how the `to_json` works for your object(s).
